### PR TITLE
[Fix] Memory leaks

### PR DESCRIPTION
**Pull request by jpfeuffer**
# Description

Especially in MzIdentML parser!

### CODE REVIEW

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/InterpolationModel.h [hunk 1]
```diff
@@ -94,6 +94,8 @@ namespace OpenMS
       interpolation_ = source.interpolation_;
       scaling_ = source.scaling_;
 
+      updateMembers_();
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/InterpolationModel.h:97]: why is this needed?

File: src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp [hunk 2]
```diff
@@ -329,7 +329,7 @@ namespace OpenMS
 
   FeatureFindingMetabo::~FeatureFindingMetabo()
   {
-
+    delete isotope_filt_svm_;
```
Comment (timosachsenberg) [inline src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp:332]: could we use a unique_ptr as well here?

File: src/openms/source/FORMAT/HANDLERS/MzMLSpectrumDecoder.cpp [hunk 3]
```diff
@@ -449,16 +449,16 @@ namespace OpenMS
...
 
-    static const XMLCh* TAG_CV = xercesc::XMLString::transcode("cvParam");
-    static const XMLCh* TAG_binary = xercesc::XMLString::transcode("binary");
-    static const XMLCh* TAG_userParam = xercesc::XMLString::transcode("userParam");
-    static const XMLCh* TAG_referenceableParamGroupRef = xercesc::XMLString::transcode("referenceableParamGroupRef");
-    static const XMLCh* TAG_accession = xercesc::XMLString::transcode("accession");
-    static const XMLCh* TAG_unit_accession = xercesc::XMLString::transcode("unitAccession");
-    static const XMLCh* TAG_value = xercesc::XMLString::transcode("value");
-    static const XMLCh* TAG_name = xercesc::XMLString::transcode("name");
+    static const XMLCh* TAG_CV = CONST_XMLCH("cvParam");
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/HANDLERS/MzMLSpectrumDecoder.cpp:454]: can you add some comment here why the macros are used?

File: src/openms/source/FORMAT/HANDLERS/MzMLSpectrumDecoder.cpp [hunk 4]
```diff
@@ -547,9 +547,11 @@ namespace OpenMS
   std::string MzMLSpectrumDecoder::domParseString_(const std::string& in, std::vector<BinaryData>& data)
   {
     // PRECONDITON is below (since we first need to do XML parsing before validating)
-    static const XMLCh* default_array_length_tag = xercesc::XMLString::transcode("defaultArrayLength");
-    static const XMLCh* id_tag = xercesc::XMLString::transcode("id");
-    static const XMLCh* binary_data_array_tag = xercesc::XMLString::transcode("binaryDataArray");
+    static constexpr XMLCh id_tag[] = {'i','d', 0};
+    static constexpr XMLCh default_array_length_tag[] = { 'd','e','f','a','u','l','t','A','r','r','a','y','L','e','n','g','t','h', 0};
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/HANDLERS/MzMLSpectrumDecoder.cpp:551]: can you add a comment why it is written like this? e.g., initializer list of uint16?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EGHTraceFitter.cpp [hunk 5]
```diff
@@ -187,6 +185,7 @@ namespace OpenMS
     this->apex_rt_ = source.apex_rt_;
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EGHTraceFitter.cpp:185]: did you check if the whole function is just the default copy assignment and we could write = default in the header? same for copy constructor

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EmgFitter1D.cpp [hunk 6]
```diff
@@ -231,7 +231,7 @@ namespace OpenMS
     {
       correlation = -1.0;
     }
-
+    delete model;
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EmgFitter1D.cpp:234]: RAII?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMRM.cpp [hunk 7]
```diff
@@ -374,7 +374,7 @@ namespace OpenMS
               std::cerr << "An error occurred during the gnuplot execution" << std::endl;
             }
           }
-
+          delete model_rt;
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMRM.cpp:377]: use RAII?

File: src/tests/class_tests/openms/source/AverageLinkage_test.cpp [hunk 8]
```diff
@@ -157,6 +157,7 @@ START_SECTION((static ClusterFunctor* create()))
 {
 	ClusterFunctor* cf = AverageLinkage::create();
   TEST_NOT_EQUAL( dynamic_cast<AverageLinkage*>(cf) , nullPointer)
+  delete cf;
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/AverageLinkage_test.cpp:160]: RAII with unique_ptr?

File: src/tests/class_tests/openms/source/BiGaussFitter1D_test.cpp [hunk 9]
```diff
@@ -122,6 +122,7 @@ START_SECTION((Fitter1D* create()))
   Fitter1D* ptr = BiGaussFitter1D::create();
   TEST_EQUAL(ptr->getName(), "BiGaussFitter1D")
 	TEST_NOT_EQUAL(ptr, nullPointer)
+	delete ptr;
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/BiGaussFitter1D_test.cpp:125]: same here?

File: src/tests/class_tests/openms/source/ControlledVocabulary_test.cpp [hunk 10]
```diff
@@ -188,6 +188,7 @@ START_SECTION(([ControlledVocabulary::CVTerm] CVTerm()))
 {
   cvterm = new ControlledVocabulary::CVTerm();
   TEST_NOT_EQUAL(cvterm, cvtermNullPointer)
+  delete cvterm;
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/ControlledVocabulary_test.cpp:191]: and here

File: src/topp/SpecLibSearcher.cpp [hunk 11]
```diff
@@ -358,7 +358,7 @@ class TOPPSpecLibSearcher :
     OPENMS_LOG_INFO << "Time needed for preprocessing data: " << (end_build_time - start_build_time) << "\n";
 
     //compare function
-    PeakSpectrumCompareFunctor* comparor = Factory<PeakSpectrumCompareFunctor>::create(compare_function);
+    PeakSpectrumCompareFunctor* comparator = Factory<PeakSpectrumCompareFunctor>::create(compare_function);
```
Comment (timosachsenberg) [inline src/topp/SpecLibSearcher.cpp:361]: wrap in unique ptr?

**Comment by timosachsenberg** [comment]
nice cleanup

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/InterpolationModel.h (hunk 1 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/InterpolationModel.h:97]: It says in the description of updateMembers that it should be at every derived copy constructor/assignment

Comment (timosachsenberg) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/InterpolationModel.h:97]: ok makes sense. One thing to watch out as it is easy to just put a = default at a copy constructor/assignment

File: src/tests/class_tests/openms/source/AverageLinkage_test.cpp (hunk 8 above)
Comment (jpfeuffer) [reply src/tests/class_tests/openms/source/AverageLinkage_test.cpp:160]: I dont think it is worth for the tests where you delete 2 lines after. Feel free to do it if you think it is important.

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EGHTraceFitter.cpp (hunk 5 above)
Comment (jpfeuffer) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EGHTraceFitter.cpp:185]: it has update_members

File: src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp (hunk 2 above)
Comment (jpfeuffer) [reply src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp]: No, needs special deleter.

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/InterpolationModel.h (hunk 1 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/InterpolationModel.h:97]: I am actually not sure why updateMembers is needed. Only makes a difference if the params of the source do not match its own members right. But this prevents = default everywhere.

**Comment by jpfeuffer** [comment]
By the way, I am giving up on SVMWrapper. It will continue leaking, especially during CV but the class is so incredibly badly designed that it is impossible to see where things get allocated. Additionally, there are two types of modes (one heap, one stack-allocated) which are only separated by if-cases in all of the methods.

File: src/openms/include/OpenMS/FILTERING/DATAREDUCTION/FeatureFindingMetabo.h [hunk 12]
```diff
@@ -311,7 +311,7 @@ e conditions are fulfilled. Mainly the
     void findLocalFeatures_(const std::vector<const MassTrace*>& candidates, double total_intensity, std::vector<FeatureHypothesis>& output_hypotheses) const;
 
     /// SVM parameters
-    svm_model* isotope_filt_svm_;
+    svm_model* isotope_filt_svm_ = nullptr;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FILTERING/DATAREDUCTION/FeatureFindingMetabo.h:314]: just FYI: you can also use {}

File: src/openms/include/OpenMS/FORMAT/HANDLERS/XMLHandler.h [hunk 13]
```diff
@@ -59,6 +60,156 @@ namespace OpenMS
   namespace Internal
   {
 
+    #define CONST_XMLCH(s) reinterpret_cast<const ::XMLCh*>(u ## s)
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/HANDLERS/XMLHandler.h:63]:
```suggestion
    // put u" in front of string literal to mark it as UTF-16 encoded, can then be directly convted to const XMLCh*
    #define CONST_XMLCH(s) reinterpret_cast<const ::XMLCh*>(u ## s)
```

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/ExtendedIsotopeFitter1D.h [hunk 14]
```diff
@@ -73,7 +73,7 @@ namespace OpenMS
     }
 
     /// return interpolation model
-    QualityType fit1d(const RawDataArrayType & range, InterpolationModel * & model) override;
+    QualityType fit1d(const RawDataArrayType & range, std::unique_ptr<InterpolationModel>& model) override;
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/ExtendedIsotopeFitter1D.h:76]: can we wrap that in pyOpenMS?

Comment (timosachsenberg) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/ExtendedIsotopeFitter1D.h:76]: and would it make sense at all because of polymorphism ???

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/InterpolationModel.h (hunk 1 above)
Comment (timosachsenberg) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/InterpolationModel.h:97]: I think we would need to check if someone sets the param class in updateMembers (which I think he should not do).

File: src/openms/source/ANALYSIS/SVM/SVMWrapper.cpp [hunk 15]
```diff
@@ -1658,6 +1656,9 @@ namespace OpenMS
       return nullptr;
     }
 
+    double temp = 0;
+    svm_problem* kernel_matrix;
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/SVM/SVMWrapper.cpp:1660]: why does this need to be a pointer? because it otherwise allocates a lot of memory on the stack?

Comment (timosachsenberg) [reply src/openms/source/ANALYSIS/SVM/SVMWrapper.cpp:1660]: svm_problem_ could also be a unique_ptr AFAIK. Copy elision will allow returning it if I understood the concept

File: src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp [hunk 16]
```diff
@@ -329,7 +304,10 @@ namespace OpenMS
 
   FeatureFindingMetabo::~FeatureFindingMetabo()
   {
-
+    if(isotope_filt_svm_ != nullptr)
```
Comment (timosachsenberg) [inline src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp:307]: coudl isotope_filt_svm_ be a unique_ptr and svm_free_and_destroy_model(&isotope_filt_svm_); be called in the lambda passed for destruction?  then this destructor could be default

File: src/openms/source/FORMAT/HANDLERS/MzIdentMLDOMHandler.cpp [hunk 17]
```diff
@@ -187,25 +188,36 @@ namespace OpenMS
...
+        //          ostringstream errBuf;
+        //          errBuf << "Error parsing file: " << message << flush;
+        OPENMS_LOG_ERROR << "XERCES error parsing file: " << message << flush << endl;
+        XMLString::release(&message);
+      }
 
-        // no need to free this pointer - owned by the parent parser object
-        xercesc::DOMDocument* xmlDoc = mzid_parser_.getDocument();
+      // we adopt/own the document so we can free it in case this DOMHandler is reused on another file.
+      xercesc::DOMDocument* xmlDoc = mzid_parser_.adoptDocument();
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/HANDLERS/MzIdentMLDOMHandler.cpp:202]: does this imply any performance / memory hit?

File: src/openms/source/FORMAT/HANDLERS/MzIdentMLDOMHandler.cpp [hunk 18]
```diff
@@ -1413,57 +1426,57 @@ namespace OpenMS
       {
         DOMElement* cl_sii = dynamic_cast<xercesc::DOMElement*>(siis->item(it->second));
         // Attributes
-        String peptide = XMLString::transcode(cl_sii->getAttribute(XMLString::transcode("peptide_ref")));
+        String peptide = StringManager::convert(cl_sii->getAttribute(CONST_XMLCH("peptide_ref")));
         peptides.push_back(peptide);
-        double exp_mz = String(XMLString::transcode(cl_sii->getAttribute(XMLString::transcode("experimentalMassToCharge")))).toDouble();
+        double exp_mz =StringManager::convert(cl_sii->getAttribute(CONST_XMLCH("experimentalMassToCharge"))).toDouble();
```
Comment (timosachsenberg) [inline src/openms/source/FORMAT/HANDLERS/MzIdentMLDOMHandler.cpp:1431]:
```suggestion
        double exp_mz = StringManager::convert(cl_sii->getAttribute(CONST_XMLCH("experimentalMassToCharge"))).toDouble();
```

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EGHTraceFitter.cpp [hunk 19]
```diff
@@ -161,10 +161,7 @@ namespace OpenMS
...
   }
 
-  EGHTraceFitter::EGHTraceFitter()
-  {
-    //setName("EGHTraceFitter");
-  }
+  EGHTraceFitter::EGHTraceFitter() = default;
 
   EGHTraceFitter::EGHTraceFitter(const EGHTraceFitter& other) :
     TraceFitter(other)
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EGHTraceFitter.cpp:167]: could potentially also be deafult

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/GaussFitter1D.cpp [hunk 20]
```diff
@@ -89,7 +89,7 @@ namespace OpenMS
 
 
     // build model
-    model = static_cast<InterpolationModel*>(Factory<BaseModel<1> >::create("GaussModel"));
+    model = std::unique_ptr<InterpolationModel>(dynamic_cast<InterpolationModel*>(Factory<BaseModel<1>>::create("GaussModel")));
```
Comment (timosachsenberg) [inline src/openms/source/TRANSFORMATIONS/FEATUREFINDER/GaussFitter1D.cpp:92]: maybe check if all these casts are needed.
I think the Factory could in principle directly return a std::unique_ptr<InterpolationModel>

File: src/tests/class_tests/openms/source/AverageLinkage_test.cpp (hunk 8 above)
Comment (timosachsenberg) [reply src/tests/class_tests/openms/source/AverageLinkage_test.cpp:160]: my argument would be: someone forgot it here :P

**Review by timosachsenberg — APPROVED** [review]

File: src/openms/source/FORMAT/HANDLERS/MzIdentMLDOMHandler.cpp (hunk 17 above)
Comment (jpfeuffer) [reply src/openms/source/FORMAT/HANDLERS/MzIdentMLDOMHandler.cpp:202]: AFAIK not. It is only a difference in ownership of the memory. Either the parser has it, or the class that calls it.

File: src/openms/source/ANALYSIS/SVM/SVMWrapper.cpp (hunk 15 above)
Comment (jpfeuffer) [reply src/openms/source/ANALYSIS/SVM/SVMWrapper.cpp:1660]: I don't know, just moved here. The svm_problems usually don't have to be pointers. But svm_models etc. you get back as pointers from libSVM. So, yes the correct way would be to use unique_ptrs for svm_model with custom deleter and do not use pointers for svm_problems until you pass it to libSVM I guess.

File: src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp (hunk 16 above)
Comment (jpfeuffer) [reply src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp:307]: yes please

Comment (jpfeuffer) [reply src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp:307]: has to be done in all of SVMWrapper though.

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/ExtendedIsotopeFitter1D.h (hunk 14 above)
Comment (jpfeuffer) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/ExtendedIsotopeFitter1D.h:76]: Of course, you can. But not with autowrap.

Comment (jpfeuffer) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/ExtendedIsotopeFitter1D.h:76]: Easy workaround is to offer a function that accepts a shared pointer since autowrap has a conversion provider for this (essentially passing the "inst" member of the already wrapped python class).

Comment (hroest) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/ExtendedIsotopeFitter1D.h:76]: or just pass a smart_ptr ?

Comment (timosachsenberg) [reply src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/ExtendedIsotopeFitter1D.h:76]: introducing a shared_ptr where a unique_ptr would suffice is bad design.
is there a possibility to wrap unique_ptr? this would be cleaner.

File: src/openms/source/ANALYSIS/SVM/SVMWrapper.cpp (hunk 15 above)
Comment (jpfeuffer) [reply src/openms/source/ANALYSIS/SVM/SVMWrapper.cpp:1660]: This shoudl be addressed in a general SVM cleanup

File: src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp (hunk 16 above)
Comment (jpfeuffer) [reply src/openms/source/FILTERING/DATAREDUCTION/FeatureFindingMetabo.cpp:307]: This should be addressed in a general SVM cleanup

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EGHTraceFitter.cpp (hunk 19 above)
Comment (jpfeuffer) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/EGHTraceFitter.cpp:167]: update members?

File: src/openms/source/TRANSFORMATIONS/FEATUREFINDER/GaussFitter1D.cpp (hunk 20 above)
Comment (jpfeuffer) [reply src/openms/source/TRANSFORMATIONS/FEATUREFINDER/GaussFitter1D.cpp]: yes they are needed
